lib/packet.cc

```diff
@@ -613,6 +613,21 @@ Packet::make(unsigned char *data, uint32_t length,
 }
 #endif
 
+/** @brief Copy the content and annotations of another packet (userlevel).
+ * @param source packet
+ * @param headroom for the new packet
+ */
+void
+Packet::copy(Packet* p, int headroom)
+{
+	_data = _head + headroom;
+	memcpy(_data,p->data(),p->length());
+	_tail = _data + p->length();
+	copy_annotations(p);
+    ptrdiff_t shift = _data - p->_data;
+    set_mac_header(p->mac_header() ? p->mac_header() + shift : 0);
+    set_network_header(p->network_header() ? p->network_header() + shift : 0, p->network_header_length());
```

Comment:
I think you missed that shift is the difference between _data and p->_data, so it should be good.